### PR TITLE
LPS-178732 LPS-176550 cause about -3.18% performance regression

### PR DESCRIPTION
**Pull request by drewbrokke**
This is an update for [LPS-178732](https://issues.liferay.com/browse/LPS-178732).

@dantewang would you be willing to review these changes? I tried to copy the pattern I saw used in `PLOEntryModelListener` for updates across the cluster. Let me know if you see any issues or ways that this can be improved. Thank you!

### CODE REVIEW

**Comment by dantewang** [comment]
@GinsonRen Please verify whether this PR fixes the performance regression. This is an very old one, so it may be hard to test directly on the original change; but we should at least see some improvement if this pull works, on master.

File: modules/apps/feature-flag/feature-flag-web/src/main/java/com/liferay/feature/flag/web/internal/manager/FeatureFlagPreferencesManager.java [hunk 1]
```diff
@@ -98,8 +112,42 @@ private PortalPreferences _getPortalPreferences(long companyId) {
 		return portalPreferencesWrapper.getPortalPreferencesImpl();
 	}
 
+	private void _notifyCluster(long companyId, String key, boolean enabled) {
+		if (!_clusterExecutor.isEnabled()) {
+			return;
+		}
+
+		MethodHandler methodHandler = new MethodHandler(
+			_onNotifyMethodKey, companyId, key, enabled);
```
Comment (dantewang) [inline modules/apps/feature-flag/feature-flag-web/src/main/java/com/liferay/feature/flag/web/internal/manager/FeatureFlagPreferencesManager.java:121]: @drewbrokke The `MethodHandler` must be provided exact arguments of the method represented by the `MethodKey`. For your `_onNotify()` method, you should provide `osgiServiceIdentifier, companyId, key, enabled`.

File: modules/apps/feature-flag/feature-flag-web/src/main/java/com/liferay/feature/flag/web/internal/manager/FeatureFlagPreferencesManager.java [hunk 2]
```diff
@@ -74,19 +79,28 @@ public Boolean isEnabled(long companyId, String key) {
...
+		_notifyCluster(companyId, key, enabled);
+	}
+
+	private static void _onNotify(
+		String osgiServiceIdentifier, long companyId, String key,
+		boolean enabled) {
+
+		FeatureFlagPreferencesManager ploEntryModelListener =
+			(FeatureFlagPreferencesManager)
+				IdentifiableOSGiServiceUtil.getIdentifiableOSGiService(
```
Comment (dantewang) [inline modules/apps/feature-flag/feature-flag-web/src/main/java/com/liferay/feature/flag/web/internal/manager/FeatureFlagPreferencesManager.java:100]: @drewbrokke In order for this util to get the service, your service must implement `IdentifiableOSGiService` interface.

File: modules/apps/feature-flag/feature-flag-web/src/main/java/com/liferay/feature/flag/web/internal/manager/FeatureFlagPreferencesManager.java (hunk 1 above)
Comment (dantewang) [reply modules/apps/feature-flag/feature-flag-web/src/main/java/com/liferay/feature/flag/web/internal/manager/FeatureFlagPreferencesManager.java:121]: @drewbrokke By the way, the code in `PLOEntryModelListener` has a bug due to BChan's SF:

```
	private static void _onNotify(
		MethodType methodType, String osgiServiceIdentifier,
		PLOEntry ploEntry) {
```
vs
```
			MethodHandler methodHandler = new MethodHandler(
				_onNotifyMethodKey, getOSGiServiceIdentifier(), methodType,
				ploEntry);
```

He sorted the method parameters, but forget the `MethodHandler`. I'll send a PR to fix it.

**Review by dantewang — APPROVED** [review]

**Comment by drewbrokke** [comment]
@dantewang I had previously sent a PR to @GinsonRen to validate the approach: https://github.com/GinsonRen/liferay-portal/pull/627#issuecomment-1562633121

It did not contain any clustering code, that is the purpose of this PR.

File: modules/apps/feature-flag/feature-flag-web/src/main/java/com/liferay/feature/flag/web/internal/manager/FeatureFlagPreferencesManager.java (hunk 1 above)
Comment (drewbrokke) [reply modules/apps/feature-flag/feature-flag-web/src/main/java/com/liferay/feature/flag/web/internal/manager/FeatureFlagPreferencesManager.java:121]: > The MethodHandler must be provided exact arguments of the method represented by the MethodKey. For your _onNotify() method, you should provide osgiServiceIdentifier, companyId, key, enabled.

Ah, good catch, thank you!

Comment (drewbrokke) [reply modules/apps/feature-flag/feature-flag-web/src/main/java/com/liferay/feature/flag/web/internal/manager/FeatureFlagPreferencesManager.java:121]: > He sorted the method parameters, but forget the MethodHandler. I'll send a PR to fix it.

Thank you Dante!

**Comment by drewbrokke** [comment]
@dantewang I just force-pushed the changes to this PR. Please let me know if you see any other issues. 

@GinsonRen is there a way to test these changes against a clustered environment as well?

**Comment by drewbrokke** [comment]
ci:test:relevant

**Comment by dantewang** [comment]
@drewbrokke Ginson's on PTO today, I'll answer that. We don't have existing support for clustered environment. We did test Liferay clustered but that was on the cloud where the cluster was built by containers.

For now, we only make sure the features on cluster work.

**Comment by dantewang** [comment]
@drewbrokke Could you send the PR to liferay-core-infra so that it can go through our review chain? Thanks!

**Comment by drewbrokke** [comment]
@dantewang yes, I will do that. Thanks!

**Comment by drewbrokke** [comment]
See https://github.com/liferay-core-infra/liferay-portal/pull/4713

**Pull request closed without merging**